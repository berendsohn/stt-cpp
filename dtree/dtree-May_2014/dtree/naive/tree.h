// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
