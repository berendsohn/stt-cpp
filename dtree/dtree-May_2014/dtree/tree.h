// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
