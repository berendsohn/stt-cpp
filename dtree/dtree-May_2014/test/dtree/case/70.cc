#define DTREE_CONFIG EndTree<WithEvert<WithEvert<WithStaticValue<dtree::DpValue<double>, WithAncValue<dtree::DpAdd<double>, WithAncAggr<dtree::Count<int>, dtree::Index_<int>, WithAncAggr<dtree::Count<int>, dtree::Index_<int>, Begin > > > > > > >
#include "test/dtree/tree.cc"
