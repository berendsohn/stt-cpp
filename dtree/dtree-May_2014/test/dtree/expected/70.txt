<<<
  1: vL34R59F0 vL-4R1F1 -> 11
  2: vL6R19F0 vL6R0F1 -> 13
  3: vL34R27F1 vL-6R3F1 -> 5
  4: vL0R-1F0 vL6R5F1 -> 13
  5: vL43R41F1 vL3R-3F0 -> 1
  6: vL0R2F1 vL-6R-6F0 -> 2
  7: vL23R13F0 vL5R-8F0 -> 3
  8: vL7R0F1 vL-3R0F0
  9: vL3R1F1 vL-4R5F0 -> 7
  10: vL1R-3F0 vL-4R-3F1 -> 12
  11: vL-1R10F0 vL-7R1F0
  12: vL-2R-4F1 vL-6R-1F1 -> 15
  13: vL6R9F1 vL-2R2F1 -> 12
  14: vL6R-4F1 vL-7R2F1 -> 5
  15: vL-2R10F1 vL-7R1F0 -> 5
>>>
<<<
 13
  \4: a1 a1 vL0R-1F0 vL6R5F1 f1 f1

  @8: a1 a1 vL7R0F1 vL-3R0F0 f0 f1

 5
      /3: a1 a1 vL11R14F1 vL-6R3F1 f1 f1
    /7: a2 a2 vL12R20F1 vL5R-8F0 f1 f1
  \9: a3 a3 vL3R1F1 vL-4R5F0 f1 f1

 12
      /2: a1 a1 vL17R6F1 vL6R0F1 f1 f1
    /6: a2 a2 vL-7R-6F0 vL-6R-6F0 f1 f0
  \13: a3 a3 vL6R9F1 vL-2R2F1 f0 f1

 5
  \14: a1 a1 vL6R-4F1 vL-7R2F1 f0 f0

    /12: a2 a2 vL-14R0F0 vL-6R-1F1 f1 f0
      \10: a1 a1 vL1R3F1 vL-4R-3F1 f1 f1
  @15: a1 a1 vL-2R10F1 vL-7R1F0 f0 f1
        /11: a1 a1 vL-35R-49F0 vL-7R1F0 f0 f1
      /1: a2 a2 vL18R-9F1 vL-4R1F1 f1 f1
    \5: a3 a3 vL31R45F0 vL3R-3F0 f0 f1
>>>
